src/badguy/granito.cpp

```diff
@@ -0,0 +1,442 @@
+        restore_original_state();
+        m_has_waved = true;
+      }
+    }
+    else
+    {
+      try_wave();
+    }
+  }
+
+  if (m_type == DEFAULT && try_jump())
+  {
+    WalkingBadguy::active_update(dt_sec);
+    return;
+  }
+
+  // Only called, when timer has finished
+  if (!m_walk_interval.started() && !m_walk_interval.check())
+  {
+    m_walk_interval.start(gameRandom.randf(1.f, 4.f));
+
+    switch (m_type)
+    {
+      case STAND:
+        if (gameRandom.rand(100) > 50)
+        {
+          // Turn around
+          m_dir = m_dir == Direction::LEFT ? Direction::RIGHT : Direction::LEFT;
+          set_action("stand", m_dir);
+        }
+
+        break;
+
+      case DEFAULT:
+      {
+        if (gameRandom.rand(100) > 50 && walk_speed == 0)
+        {
+          // Turn around
+          m_dir = m_dir == Direction::LEFT ? Direction::RIGHT : Direction::LEFT;
+          set_action("stand", m_dir);
+        }
+        else
+        {
+          // Walk/stop
+          if (walk_speed > 0)
+          {
+            walk_speed = 0;
+            m_state = STATE_STAND;
+            m_original_state = STATE_STAND;
+            m_physic.set_velocity_x(0);
+            set_action("stand", m_dir);
+          }
+          else
+          {
+            m_dir = (gameRandom.rand(2) == 0 ? Direction::LEFT : Direction::RIGHT);
+            walk_speed = 80;
+            m_state = STATE_WALK;
+            m_original_state = STATE_WALK;
+            m_physic.set_velocity_x(80 * (m_dir == Direction::LEFT ? -1 : 1));
+            set_action(m_dir);
+          }
+        }
+
+        break;
+      }
+
+      default:
+        break;
+    }
+  }
+
+  WalkingBadguy::active_update(dt_sec);
+
+  m_stepped_on = false;
+}
+
+HitResponse
+Granito::collision_player(Player& player, const CollisionHit& hit)
+{
+  if (m_state == STATE_SIT || m_type == WALK) return FORCE_MOVE;
+
+  if (hit.top)
+  {
+    m_stepped_on = true;
+
+    if (m_state != STATE_LOOKUP)
+    {
+      m_state = STATE_LOOKUP;
+      walk_speed = 0;
+      m_physic.set_velocity_x(0);
+      set_action("lookup", m_dir);
+
+      // Don't wave again because we've already spotted the player
+      m_has_waved = true;
+    }
+  }
+
+  return FORCE_MOVE;
+}
+
+HitResponse
+Granito::collision(GameObject& other, const CollisionHit& hit)
+{
+  if (hit.top)
+    m_col.propagate_movement(m_col.get_movement());
+
+  if (hit.bottom)
+  {
+    if (m_state == STATE_SIT) return WalkingBadguy::collision(other, hit);
+
+    // yo big granito can i sit on top of your head?
+    GranitoBig* granito = dynamic_cast<GranitoBig*>(&other);
+
+    if (!granito)
+    {
+      // i'm not a granito
+      return WalkingBadguy::collision(other, hit);
+    }
+
+    if (granito->m_carrying != nullptr)
+    {
+      // sorry im already carrying this guy
+      return WalkingBadguy::collision(other, hit);
+    }
+
+    // sure dude
+    granito->m_carrying = this;
+
+    // yay
+    m_state = STATE_SIT;
```

Comment:
Comments are fine, could follow the convention of uppercase first letter though (and preferably, an end-of-sentence mark).